website/static/js/nodesPrivacy.js

```diff
@@ -0,0 +1,280 @@
+    self.nodesChangedPrivate = ko.observableArray([]);
+    self.hasChildren = ko.observable(false);
+    $('#nodesPrivacy').on('hidden.bs.modal', function () {
+        self.clear();
+    });
+
+    /**
+     * get node tree for treebeard from API V1
+     */
+    self.fetchNodeTree = function() {
+        return $.ajax({
+            url: treebeardUrl,
+            type: 'GET',
+            dataType: 'json'
+        }).done(function(response) {
+            var i = 0;
+            self.nodesOriginal = getNodesOriginal(response[0], self.nodesOriginal);
+            Object.size = function(obj) {
+                var size = 0, key;
+                for (key in obj) {
+                    if (obj.hasOwnProperty(key)) size++;
+                }
+                return size;
+            };
+            // Get the size of an object
+            var size = Object.size(self.nodesOriginal);
+            self.hasChildren(size > 1);
+            var nodesState = $.extend(true, {}, self.nodesOriginal);
+            var nodeParent = response[0].node.id;
+            //change node state and response to reflect button push by user on project page (make public | make private)
+            nodesState[nodeParent].public = response[0].node.is_public = !parentIsPublic;
+            nodesState[nodeParent].changed = true;
+            self.nodesState(nodesState);
+        }).fail(function(xhr, status, error) {
+            $osf.growl('Error', 'Unable to retrieve project settings');
+            Raven.captureMessage('Could not GET project settings.', {
+                url: treebeardUrl, status: status, error: error
+            });
+        });
+    };
+
+    self.page = ko.observable(self.WARNING);
+
+    self.pageTitle = ko.computed(function() {
+        return {
+            warning: 'Warning',
+            select: 'Change privacy settings',
+            confirm: 'Projects and components affected'
+        }[self.page()];
+    });
+
+    self.message = ko.computed(function() {
+        return {
+            warning: MESSAGES.makeProjectPublicWarning,
+            select: MESSAGES.selectNodes,
+            confirm: MESSAGES.confirmWarning
+        }[self.page()];
+    });
+
+    self.selectProjects =  function() {
+        self.page(self.SELECT);
+    };
+
+    self.confirmWarning =  function() {
+        var nodesState = ko.toJS(self.nodesState);
+        for (var node in nodesState) {
+            if (nodesState[node].changed) {
+                if (nodesState[node].public) {
+                    self.nodesChangedPublic().push(nodesState[node].title);
+                }
+                else {
+                    self.nodesChangedPrivate().push(nodesState[node].title);
+                }
+            }
+        }
+        self.page(self.CONFIRM);
+    };
+
+    self.confirmChanges =  function() {
+        var nodesState = ko.toJS(self.nodesState());
+        //patchNodesPrivacy(nodesState);
+        nodesState = Object.keys(nodesState).map(function(key) {
```

Comment:
Minor: It's a bit misleading to overwrite the `nodesState` variable here because it isn't the same as `self.nodesState()`. How about declaring this as a separate variable, e.g. `nodesStateValues`.